tasks/discover-granules/tests/index.js

```diff
@@ -116,60 +115,150 @@ test('discover granules using HTTP', async (t) => {
-    'granule-2.nc', 'granule-2.nc.md5',
-    'granule-3.nc', 'granule-3.nc.md5'
-  ];
+const discoverGranulesUsingS3 = (configure, assert = assertDiscoveredGranules) =>
+  async (t) => {
+    const { event, event: { config } } = t.context;
+    // State sample files
+    const files = [
+      'granule-1.nc', 'granule-1.nc.md5',
+      'granule-2.nc', 'granule-2.nc.md5',
+      'granule-3.nc', 'granule-3.nc.md5'
+    ];
 
-  config.sourceBucketName = randomString();
-  config.collection.provider_path = randomString();
+    config.sourceBucketName = randomString();
+    config.collection.provider_path = randomString();
 
-  configure(config);
+    configure(t);
 
-  await validateConfig(t, config);
-  await s3().createBucket({ Bucket: config.sourceBucketName }).promise();
+    validateConfig(t, config);
+    await s3().createBucket({ Bucket: config.sourceBucketName }).promise();
 
-  try {
-    await Promise.all(files.map((file) =>
-      s3().putObject({
-        Bucket: config.sourceBucketName,
-        Key: `${config.collection.provider_path}/${file}`,
-        Body: `This is ${file}`
-      }).promise()));
-    await assertDiscoveredGranules(t, await discoverGranules(event));
-  } finally {
-    // Clean up
-    await recursivelyDeleteS3Bucket(config.sourceBucketName);
-  }
-};
+    try {
+      await Promise.all(files.map((file) =>
+        s3().putObject({
+          Bucket: config.sourceBucketName,
+          Key: `${config.collection.provider_path}/${file}`,
+          Body: `This is ${file}`
+        }).promise()));
+      assert(t, await discoverGranules(event));
+    } finally {
+      // Clean up
+      await recursivelyDeleteS3Bucket(config.sourceBucketName);
+    }
+  };
 
 test('discover granules using S3',
-  discoverGranulesUsingS3((config) => {
+  discoverGranulesUsingS3(({ context: { event: { config } } }) => {
+    config.provider = {
+      id: 'MODAPS',
+      protocol: 's3',
+      host: config.sourceBucketName
+    };
+  }));
+
+test('discover granules without collection files config using S3',
+  discoverGranulesUsingS3(({ context: { event: { config } } }) => {
+    // Without files config we should still discover granules, but the
+    // discovered granules will have empty files arrays.
+    config.collection.files = [];
+    config.provider = {
+      id: 'MODAPS',
+      protocol: 's3',
+      host: config.sourceBucketName
+    };
+  }, (t, output) => {
+    validateOutput(t, output);
+    t.is(output.granules.length, 3);
+    output.granules.forEach(({ files }) => t.is(files.length, 0));
+  }));
+
+test('discover granules without collection files config, but configuring collection to ignore it, using S3',
+  discoverGranulesUsingS3(({ context: { event: { config } } }) => {
+    // Without files config we should still discover granules, and the
+    // discovered granules' files arrays will include all files because we're
+    // ignoring the (empty) files config for filtering files.
```

Comment:
Where does `granuleIdExtraction` get set so that the files are grouped into 3 granules?

Reply:
This tests uses `tasks/discover-granules/tests/fixtures/mur.json`, which includes a value for `granuleIdExtraction`, and the files for the S3 tests are defined on lines 133-137, where you'll find 3 files ending in `.nc` and 3 files ending in `.md5`, giving 3 granules, each containing 1 `.nc` file and `.md5` file.